### ISSUE TITLE
State of Ceres?

### ISSUE THREAD

**Issue by Civil** [graphite-project/carbon#361]
I know that it's not the best place to have a conversation about state of the Ceres, but it seems that issues in Ceres repo are ignored for some time. So I'm opening it here, to discuss a few things about Ceres.

As you know, there are a lot of problems in current Ceres repo (and same for ceres-related utilites in megacarbon branch). For quite some time I'm using @dkulikovsky fork in production at my current work. It works more or less OK, but... I've also made some changes to fix some issues (like rewrite rollup process from scratch, fix several bugs in ceres.py that were resulting in forsaken files at rollup and so on). And in fact I'd really like to see all those changes in mainline (also I'd like to try to get changes from  @dkulikovsky to the mainline repo).

So what's the state of the ceres right now? Is it abandoned completely or there are some chances to get some of PRs to it? And another question - if I'll get permissions to do that, will it be possible to at least discuss all the changes that were made by @dkulikovsky and me to ceres, rollup and merge?

**Comment by ccope** [comment]
There was some recent progress in https://github.com/graphite-project/carbon/pull/358 which makes it sound like support for ceres in the carbon master branch might happen in the near future. After that, I bet there will be more interest in working on ceres.

**Comment by Civil** [comment]
Yeah, I also hope for that. Though I've opened this issue to have a dialog about some patches for carbon and ceres. There are a lot of things to do and maybe I can offer some help.

**Comment by SEJeff** [comment]
@ccope Pretty much. We are all volunteers and sometimes it is a bit difficult. I'd be a good bit more confident in merging some of the changes to ceres if it had more/better tests. It sucks that there aren't a ton as is, but you have to start somewhere!

Ceres has some warts, but it isn't dead. I think we'll have a bit more time to look into ceres post 0.9.13 release of all of the other graphite components.

**Comment by Civil** [comment]
@SEJeff the problem is, that from my point of view, mainline ceres is unusable. @dkulikovsky done a lot of work to get it more or less usable, though I think none of his patches got tests. I've done some work recently (as PRs to his repos), though also no tests for that (for some, maybe, I can provide test cases, for other it'll take too much time to get those things reproducible, especially for everything related to rollup). Also coding style of ceres... it's basicly incompatible with python 3.x (easy to fix), quite strange (using python's keywords as a variables, and so on) and a lot more stuff to do. The project maybe not dead, but it's unusable right now, there were no work at all for 2 years in the repos, there are several PRs without any comment and so on, and so on.

So basically the question is - do you need some help with merging some of patches (and do you need those patches) or not? Especially if most of the things will be without any test cases.

**Comment by justanyone** [comment]
We're using Ceres (at a large online retailer based in or near Hoffman
Estates, IL that rhymes with Beers FoeBuck) handling ... <clearing throat
noise> several million metrics per minute.  We've found it does have
limitations, but works for us in general.

We're in the process of switching to use MongoDB as a backend storage
engine, but in the meantime, Ceres has functioned, and is functioning
reasonably well for us.

I'm not sure if we've made mods to the maintenance routines that aren't
committed back up to the project.  I'll find out from my coworker about
that and clone/branch/commit/push if I can get a spare cycle.

We're not really interested in enhancing it greatly anymore since we're
moving away from it, but it's functioned quite well for at least 2 years
here.

-- Kevin J. Rice

**Comment by Civil** [comment]
> Ceres has functioned, and is functioning
reasonably well for us.

Get a look at a fork by dkulikovsky (https://github.com/dkulikovsky/ceres) and in particularly commits he've made there. There were some problems like memory leaks inside ceres iteself. There were some other minor stuff. I've also found that in his fork there are some other problems, like 'deleteBefore' was a bit broken and in some cases it produces malformed data, leaving one-point-files behind and some other things, like it interprets MAX_SLICE_GAP in bytes, but in config it's described as it is points. Also it have some problems when you got overlapping slices. For example:

you've got two slices, one for 1420562100-1420907700 time range with retention of 300s (1152 points), and another for 1420821300-now with 60s. If you'll try to read the data from 1420562100 till 1420907700, you'll expect to receive 1152 points, but you'll receive 1440 points instead, though timeStep will be set to "300" (it'll recalculate data for the interval 1420821300-1420907700 from slice with 60s retention, but it will, instead of merging the data, append that to the previous results from slice with 300s retention). I know that it shouldn't happen in normal situations, but original rollup+merge plugins can produce such results.

So from my point of view, all those small bugs there makes Ceres unreliable - you can't trust the data you'll receive from mainline ceres, and that's why it need some love in getting it into normal shape. Most of the time it works more or less ok, but there are some cases where rollup or merge can break historical data, or produce output that you and your software expects. The example from above, about 'ceresNode.read' returning more data than you'd expect, results in malformed output if you've got only parts of the data in one slice and parts in another - actual result will be shifted. It also will make you harder to write some additional scripts that'll do some work with ceres (I've found that issue when I was creating a script that should produce summed values for several metrics, to populate aggregated values with historical data)

**Comment by obfuscurity** [comment]
Ok, so what will it take to get some of those fixes submitted as PRs upstream?

**Comment by Civil** [comment]
@obfuscurity as far as I know, @dkulikovskiy tried to get his fixes as PRs, but when he saw that there are no actions and no discussions about them here for some time (half a year or so) he stopped sending it. My fixes are also against his branch, so to get them here, I need to also push his changes.

I see no really big deal in that, if there'll be somebody in upstream, interested in discussing those things. And I think I can get permission from the @dkulikovsky to try to push his patches here (get his commits as a basis and fix any issues you'll found during PRs discussion).

**Comment by Civil** [comment]
I've opened issues for all stuff I've found by myself, with links to commits that should fix that. Tough it's not a PR requests...

**Comment by dkulikovskiy** [comment]
Vladimir,

I'm not dkulikovsky, be careful. My login is dkulikovskiy ( i before y in
login)

понедельник, 19 января 2015 г. пользователь Vladimir Smirnov написал:

> I've opened issues for all stuff I've found by myself, with links to
> commits that should fix that. Tough it's not a PR requests...
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/graphite-project/carbon/issues/361#issuecomment-70477472>
> .
>


-- 
Best regards,
Daniil Kulikovskiy

Network engineering services
tel.:+7 963 7101653
